pandas_datareader/data.py

```diff
@@ -581,42 +582,105 @@ def fetch_data(url, name):
-        dataset = [d.split() for d in data[left_edge:right_edge]]
-        if len(dataset) > 10:
-            ncol_raw = np.array(lmap(len, dataset))
-            ncol = np.median(ncol_raw)
-            header_index = np.where(ncol_raw == ncol - 1)[0][-1]
-            header = dataset[header_index]
-            ds_header = dataset[header_index + 1:]
-            # to ensure the header is unique
-            header = ['{0} {1}'.format(j, hj) for j, hj in enumerate(header,
-                                                                     start=1)]
-            index = np.array([d[0] for d in ds_header], dtype=int)
-            dataset = np.array([d[1:] for d in ds_header], dtype=float)
-            datasets[i] = DataFrame(dataset, index, columns=header)
+            data = zf.open(zf.namelist()[0]).read().decode()
+
+    return data
+
+
+def _parse_date_famafrench(x):
+    # what's the correct python way to do that ??
+    x = x.strip()
+    try: return dt.datetime.strptime(x, '%Y')
+    except: pass
+    try: return dt.datetime.strptime(x, '%Y%m')
+    except: pass
+    return to_datetime(x)
+
+
+def get_data_famafrench(name):
+    """
+    Get data for the given name from the Fama/French data library.
+
+    For annual and monthly data, index is a pandas.PeriodIndex, otherwise
+    it's a pandas.DatetimeIndex.
+
+    Returns
+    -------
+    df : a dictionary of pandas.DataFrame. Tables are accessed by integer keys.
+         See df['DESCR'] for a description of the dataset
+    """
+    params = {'index_col': 0,
+              'parse_dates': [0],
+              'date_parser': _parse_date_famafrench}
+
+    # headers in these files are not valid
+    if name.endswith('_Breakpoints'):
+        c = ['<=0', '>0'] if name.find('-') > -1 else ['Count']
+        r = range(0, 105, 5)
+        params['names'] = ['Date'] + c + list(zip(r, r[1:]))
+        params['skiprows'] = 1 if name != 'Prior_2-12_Breakpoints' else 3
+
+    doc_chunks, tables = [], []
+    data = _download_data_famafrench(name)
+    for chunk in data.split(2 * '\r\n'):
+        if len(chunk) < 800:
+            doc_chunks.append(chunk.replace('\r\n', ' ').strip())
+        else:
+            tables.append(chunk)
+
+    datasets, table_desc = {}, []
+    for i, src in enumerate(tables):
+        match = re.search('^\s*,', src, re.M)  # the table starts there
+        start = 0 if not match else match.start()
+
+        df = read_csv(StringIO('Date' + src[start:]), **params)
+        try: df = df.to_period()
+        except: pass
+        datasets[i] = df
+
+        title = src[:start].replace('\r\n', ' ').strip()
+        shape = '({0} rows x {1} cols)'.format(*df.shape)
+        table_desc.append('{0} {1}'.format(title, shape).strip())
+
+    descr = '{0}\n{1}\n\n'.format(name.replace('_', ' '), len(name) * '-')
+    if doc_chunks: descr += ' '.join(doc_chunks).replace(2 * ' ', ' ') + '\n\n'
+
+    table_descr = map(lambda x: '{0:3} : {1}'.format(*x), enumerate(table_desc))
```

Comment:
any suggestion for a nicer simpler formatting ?

IMO, `datasets` should also implement `__getindex__` with table names instead in addition to integers, the table names should also be accessible in a list and `datasets['DESCR']` should read `datasets.DESCR`.